mypy/semanal.py

```diff
@@ -1064,17 +1064,26 @@ def setup_self_type(self) -> None:
         assert self.type is not None
         info = self.type
         if info.self_type is not None:
-            if has_placeholder(info.self_type.upper_bound):
+            if has_placeholder(info.self_type.upper_bound) or has_placeholder(
+                info.self_type.default
+            ):
                 # Similar to regular (user defined) type variables.
                 self.process_placeholder(
                     None,
-                    "Self upper bound",
+                    "Self upper bound or default",
```

Comment:
Maybe better to have a separate message for the two cases?

Reply:
In this particular instance I believe it isn't even necessary to check it at all. For `Self` the `default` is always set to `AnyType(TypeOfAny.from_omitted_generics)` and thus can't include a placeholder. I've split the calls in `process_typevar_declaration` though.

Also noticed that I was missing these checks in `process_paramspec_declaration` and `process_typevartuple_declaration` which I added now.